### ISSUE TITLE
[feature request] Querying by tag across measurements should exclude measurements without the tag, not fail with an error

### ISSUE THREAD

**Issue by erikogenvik** [influxdata/influxdb#3806]
I'm not sure if the behaviour I'm describing is a bug or intended behaviour; I couldn't get a clear consensus from the docs.

The problem is that when doing a query using a regexp for the series and matching against a tag, the query will fail if any of the matching series is missing the tag.
My expectation would be to instead return those series where there were a matching tag, and just ignoring those that didn't have the tag.

My use case for these kind of queries is when doing cleanup of the db. We have a couple of series with various names. Some of the series are attached to "assets", and thus have an "asset" tag. When one of these assets are removed from the system we also want to prune the data and remove all time series that have a matching "asset" tag.

Another issue with this behaviour is that it's too easy to write queries which works one day, and then fails the next day because a new separate time series had been added, missing a tag.
I'll demonstrate with an example:
```
curl -G http://localhost:8086/query --data-urlencode "q=CREATE DATABASE tagtest"
curl -i -XPOST 'http://localhost:8086/write?db=tagtest' --data-binary 'test_point,tag1=one value=1'
curl -G http://localhost:8086/query --data-urlencode "db=tagtest"  --data-urlencode "q=SELECT * FROM /.*/ where tag1='one'"
```
The last query for all series with the a tag of name "tag1" and value "1" will succeed.
However, if another time series is added without the tag, and the same query is run again it will fail:
```
curl -i -XPOST 'http://localhost:8086/write?db=tagtest' --data-binary 'test_point_without_tag value=1'
curl -G http://localhost:8086/query --data-urlencode "db=tagtest"  --data-urlencode "q=SELECT * FROM /.*/ where tag1='one'"
```

**Comment by mjdesa** [comment]
@erikogenvik what version of influxdb are you using?

**Comment by beckettsean** [comment]
@erikogenvik in general querying across measurements is not advisable in InfluxDB 0.9. The regex will match multiple measurements, but if those measurements don't share the same fields it can cause problems. Is there a reason you have many measurements? It was common in InfluxDB 0.8 to have thousands of series, but in InfluxDB 0.9, measurements should be in the dozens or hundreds. Each measurement can contain millions of series, however.

> However, if another time series is added without the tag, and the same query is run again it will fail

Can you clarify? Do you get no results at all for the query? Do you get an empty result set?

**Comment by erikogenvik** [comment]
@mjdesa I'm using 0.9.2

@beckettsean Sorry, I should have presented the output. This is what I get when there's only a single serie with a matching tag
```
curl -G http://localhost:8096/query --data-urlencode "q=CREATE DATABASE tagtest"
{"results":[{}]}

curl -i -XPOST 'http://localhost:8096/write?db=tagtest' --data-binary 'test_point,tag1=one value=1'
HTTP/1.1 204 No Content
Request-Id: 40dededf-4afa-11e5-b185-000000000000
X-Influxdb-Version: 0.9.2.1
Date: Tue, 25 Aug 2015 07:24:00 GMT

curl -G http://localhost:8096/query --data-urlencode "db=tagtest"  --data-urlencode "q=SELECT * FROM /.*/ where tag1='one'"
{"results":[{"series":[{"name":"test_point","tags":{"tag1":"one"},"columns":["time","value"],"values":[["2015-08-25T07:23:59.532589749Z",1]]}]}]}
```

And then when adding a new serie without the tag I get an error result back instead.
```
curl -i -XPOST 'http://localhost:8096/write?db=tagtest' --data-binary 'test_point_without_tag value=1'
HTTP/1.1 204 No Content
Request-Id: b411f56d-4afa-11e5-b187-000000000000
X-Influxdb-Version: 0.9.2.1
Date: Tue, 25 Aug 2015 07:27:12 GMT

curl -G http://localhost:8096/query --data-urlencode "db=tagtest"  --data-urlencode "q=SELECT * FROM /.*/ where tag1='one'"
{"results":[{"error":"unknown field or tag name in where clause: tag1"}]}
```

My expectation is to instead get the same result back as the first time the query was run, with those series that are missing the tag being ignored.
I'm concerned that this behaviour makes any application using regexp to query for series very brittle, since there's now an implicit dependency. If ever a new series matching the regexp is added to the database, but is missing a tag, the query will fail.

Considering the type of queries you want to do with regexp they are often "meta" queries. I.e. "show all of the series belonging to a specific asset so that the user then can drill down to one of them", or as in my case "get all series belonging to an asset so that they can be removed".

**Comment by beckettsean** [comment]
> I'm concerned that this behaviour makes any application using regexp to query for series very brittle, since there's now an implicit dependency. If ever a new series matching the regexp is added to the database, but is missing a tag, the query will fail.

> Considering the type of queries you want to do with regexp they are often "meta" queries. I.e. "show all of the series belonging to a specific asset so that the user then can drill down to one of them", or as in my case "get all series belonging to an asset so that they can be removed".

I understand your concern, @erikogenvik, but your schema should be designed so that there are only a few measurements (the thing in the FROM clause) and queries almost always operate within a single measurement. If you have thousands of measurements that points to an incorrect or inefficient schema. Inside a measurement all series are automatically merged, and only filtered out by the use of tags. 

Your current use case is clearly just a test, so I can't suggest how to alter your schema to fit with the current design guidelines. Can you share a few example queries that you would really want to run?

Said another way, measurements are independent of each other, and querying across them is both difficult and not a current design goal. Series within a measurement are merged by default and the tags and fields are a shared concept. It's not yet pushed to the main site, but see https://new.influxdb.com/docs/v0.9/concepts/08_vs_09.html for more detail.

**Comment by erikogenvik** [comment]
It sounds like we're talking about different things. The issue here, as you correctly noted when renaming the issue, is that measurements without the tag should be excluded in the query.
That issue still stands even if the number of measurements in the schema only number two, or any other small number.

My specific use case is that I want to remove those measurements that are tagged with a specific tag. My first approach was to use "SELECT * from ... LIMIT 1" to get those, and then walk through them and delete them. This is how I ran into this issue.

But a more preferable approach would probably be to use "SHOW MEASUREMENTS" instead, with a "where" statement.
Or even do a "DROP SERIES where...", which brought up another bug (I'll post separately).

**Comment by beckettsean** [comment]
@erikogenvik Given your goal of finding measurements with a particular tag, you're correct that `SHOW MEASUREMENTS` is a better tool, and it does support a `WHERE` clause:

`SHOW MEASUREMENTS WHERE tag1 = 'one'` will return all measurements with `tag1` set to "one".

**Comment by ckmaresca** [comment]
Querying for tags present in measurements would be a very useful feature.   Something like `SELECT * WHERE [tag_name] = [tag_value]`

The use case I would point to is a measurement where one of the tags identifies a specific source or region and you are trying to get an overview of what is going on for that source/region.  You could recursively query to get all the data (using `SHOW MEASUREMENTS` but it would be way, way easier if there was a shortcut that returned actual data....

**Comment by beckettsean** [comment]
@ckmaresca I don't follow what you are saying. `SELECT * FROM measurement WHERE tag_key = tag_value` is already support syntax, and could be done against all measurements with `SELECT * FROM /.*/ WHERE tag_key = tag_value`

**Comment by ckmaresca** [comment]
Ah, my mistake - didn't see it anywhere in the docs (or maybe I just missed it).   From this question, I figured it was just not supported.  Thanks for the clarification.

**Comment by beckettsean** [comment]
@ckmaresca the query works but only if all measurements matching the FROM clause have the same tag keys. Thus the title of this issue, that the query should simply exclude measurements matching the FROM clause that lack the WHERE clause tag key, not throw an error.

**Comment by jsternberg** [comment]
I think this now works correctly with the query engine refactor that will be released soon with 0.11.

```
> insert cpu1,host=server01 value=3
> insert cpu2 value=4
> select * from cpu1, cpu2 where host = 'server01'
name: cpu1
----------
time                    host            value
1458046800621338740     server01        3
```

Please reopen if I got something wrong or this still doesn't work. Thanks!
